posts/inside-rust/2023-05-02-stabilizing-async-fn-in-trait.md

```diff
@@ -0,0 +1,197 @@
+
+The easiest way to explain what we are going to stabilize is to use a code example. To start, we will permit the use of `async fn` in traits definitions...
+
+```rust
+trait HealthCheck {
+    async fn check(&mut self) -> bool;
+}
+```
+
+...and you can then use `async fn` in the corresponding impl:
+
+```rust
+impl HealthCheck for MyHealthChecker {
+    async fn check(&mut self) -> bool {
+        do_async_op().await
+    }
+}
+```
+
+Traits with async functions can then be used as you normally would:
+
+```rust
+async fn do_health_check(hc: impl HealthCheck) {
+    if !hc.check().await {
+        log_health_check_failure().await;
+    }
+}
+```
+
+**Status:** This functionality was described in [RFC 3185], merged on Dec 7, 2021, and was covered in detail in our [previous blog post][pp].
+
+[RFC 3185]: https://rust-lang.github.io/rfcs/3185-static-async-fn-in-trait.html
+
+
+## MVP Part 2: Send bounds and associated return types
+
+There is one complication that arises when using async functions in traits that doesn't arise with sync functions. Many async runtimes -- notably including the default configurations of [Tokio] and [async-std] -- use a workstealing thread scheduler. This means that futures may move between worker threads dynamically to achieve load balancing. As a result, the future must only capture `Send` data.
+
+[Tokio]: https://tokio.rs/
+
+[async-std]: https://async.rs/
+
+If you author a generic async function that spawns tasks on one of those runtimes, however, you will start to get compilation errors ([playground](XXX)):
+
+```rust
+async fn do_health_check_par(hc: impl HealthCheck) {
+    tokio::task::spawn(async move {
+        if !hc.check().await {
+            log_health_check_failure().await;
+        }
+    });
+}
+```
+
+The problem is that the future returned by `hc.check()` isn't guaranteed to be `Send`. It might access non-Send data. The solution is to add a `Send` bound, but given that this is an async function, it's not obvious how to do that. How do we talk about the future returned by a call to `hc.check()`? Associated return types provide the answer. We can convert the above function to use an explicit type parameter `HC` (instead of `impl HealthCheck`) and then add a new bound, `HC::check(): Send`. This says "the value returned by `HC::check` must be of `Send` type":
+
+```rust
+async fn do_health_check_par<HC>(hc: HC)
+where
+    HC: HealthCheck + Send + 'static,
+    HC::check(): Send, // <-- associated return type
+{
+    tokio::task::spawn(async move {
+        if !hc.check().await {
+            log_health_check_failure().await;
+        }
+    });
+}
+```
+
+Of course, it's kind of unfortunate that we had to rewrite from taking an `impl HealthCheck` to an explicit `HC` type parameter in order to use this notation. RFC #2289, "associated type bounds", introduced a compact notation to address this problem. That RFC is not part of this MVP, but if it were stabilized, then one could simply write:
```

Comment:
Not sure whether the link was intentionally omitted, but perhaps worth including if not?

```suggestion
Of course, it's kind of unfortunate that we had to rewrite from taking an `impl HealthCheck` to an explicit `HC` type parameter in order to use this notation. [RFC 2289](https://github.com/rust-lang/rfcs/pull/2289), "associated type bounds", introduced a compact notation to address this problem. That RFC is not part of this MVP, but if it were stabilized, then one could simply write:
```